src/visitor/RewriteAllSubExprVisitor.cpp

```diff
@@ -0,0 +1,121 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "visitor/RewriteAllSubExprVisitor.h"
+
+namespace nebula {
+namespace graph {
+
+RewriteAllSubExprVisitor::RewriteAllSubExprVisitor(
+                          const Expression* rewrittenSubExpr,
+                          const Expression* rewriteSubExpr,
+                          Expression* originExpr)
+    : rewrittenSubExpr_(rewrittenSubExpr),
+      rewriteSubExpr_(rewriteSubExpr),
+      originExpr_(originExpr) {
+    DCHECK(!!rewrittenSubExpr);
+    DCHECK(!!rewriteSubExpr);
+    DCHECK(!!originExpr);
+    needRewriteRoot_ = *rewrittenSubExpr_ == *originExpr_;
+}
+
+void RewriteAllSubExprVisitor::visit(TypeCastingExpression *expr) {
+    if (needRewriteRoot_) return;
+    auto* operand = expr->operand();
+    if (*operand == *rewrittenSubExpr_) {
```

Comment:
It's really confused to compare two expression objects directly? It's dangerous for current expression implementation which use many pointers in it if one expression doesn't overwrite the `operator==`. I feel you should rethink how to rewrite or replace some subexpression by an existing expression.

Reply:
ACK. I implemented a more abstract visitor in  [PR#870](https://github.com/vesoft-inc/nebula-graph/pull/870) , so the visitor here will be abandoned.

Reply:
Already delete this visitor.